File: src/org/lineageos/xiaomi_bluetooth/settings/configs/SerialNumberController.java
package org.lineageos.xiaomi_bluetooth.settings.configs;

import android.content.Context;

import androidx.annotation.NonNull;

import org.lineageos.xiaomi_bluetooth.EarbudsConstants;

import java.nio.charset.StandardCharsets;


public class SerialNumberController extends ConfigController {

    public SerialNumberController(@NonNull Context context, @NonNull String preferenceKey) {
        super(context, preferenceKey);
    }

    @Override
    public int getConfigId() {
        return EarbudsConstants.XIAOMI_MMA_CONFIG_SN;
    }

    @Override
    public int getExpectedConfigLength() {
        return 20;
    }

    @Override
    public String getSummary() {
        if (getConfigValue() == null) {
            return null;
        }

        try {
            return new String(getConfigValue(), StandardCharsets.UTF_8);
        } catch (Exception e) {
            return null;
        }
    }

}

File: src/org/lineageos/xiaomi_bluetooth/configs/SerialNumberController.kt
package org.lineageos.xiaomi_bluetooth.configs

import android.content.Context
import org.lineageos.xiaomi_bluetooth.EarbudsConstants.XIAOMI_MMA_CONFIG_SN
import java.nio.charset.StandardCharsets

class SerialNumberController(
    context: Context,
    preferenceKey: String
) : ConfigController(context, preferenceKey) {

    override val configId = XIAOMI_MMA_CONFIG_SN
    override val expectedConfigLength = 20

    override val summary: String?
        get() = configValue?.let {
            runCatching { String(it, StandardCharsets.UTF_8) }.getOrNull()
        }

}
